### DISCUSSION TITLE
copy.deepcopy when inserting elements in a ring

### DISCUSSION THREAD

**Discussion by simoneliuzzo** [atcollab/at#477] [Python AT]
I have spent several hours tracking back this issue. Thanks a lot to @lcarver, who spotted the problem.

I tried to insert elements in the lattice, thin correctors. 

I forgot to add copy.deepcopy(element) so all elements were being updated as a family in later computations. 

I think I am not the only one that fell into this trap. Below a code snippet:

```
thin_cor = at.ThinMultipole('ThinCorrector', [0], [0], max_order=0)
ring_thin = copy.deepcopy(ring)

for c, ic in enumerate(ind_cor):
    ring_thin.insert(ic+c, thin_cor)     # <<-- FAILS TO DO WHAT IS GENERALLY INTENDED, adds a family of magnets 
    ring_thin.insert(ic+c, copy.deepcopy(thin_cor))  # <<-- DOES WHAT IS GENERALLY INTENDED, adds independent elements

```

I find that in pyAT the use of copy.deepcopy() is necessary very often in order. It is in fact a really frequent reason for unreasonable results in my code. 

Is there a way to make more clear to the user that an element is not an individual but it is linked to many others?
Is there a way to 'un-link' all magnets in a lattice? (`independent_ring = [copy.deepcopy(el) for el in ring]`)

**Comment by lfarv** [comment]
No, you should not have to use `deep copy() `so often. For instance, to add individual elements, you would write:
```
for c, ic in enumerate(ind_cor):
    ring_thin.insert(ic+c, at.ThinMultipole('ThinCorrector', [0], [0], max_order=0))
```
Keep in mind that in python **everything** is a reference.
Also, `deep_copy` is costly: it has to investigate recursively there contents of an object. Creating a new one is more efficient !
